var/spack/repos/builtin/packages/libuuid/package.py

```diff
@@ -20,6 +22,7 @@ class Libuuid(AutotoolsPackage, SourceforgePackage):
 
     depends_on("c", type="build")  # generated
 
-    provides("uuid")
+    if sys.platform not in ["darwin", "windows"]:
```

Comment:
Would it be possible to do

```
for x in ('linux', 'cray'):
  with when("platform=x"):
    ...
```

?

while similar in terms of handling `linux, darwin, cray, windows`, this keys off of the spec, vs. the platform context we are in.

Reply:
Do we really need a for-loop? It's 2 lines of code either way. I find it simpler to skip the for-loop.

Reply:
I think @scheibelp's point, if I understand it, is that instead of using the host's platform (via `sys`) we use the platform as defined by the spec. I am not familiar enough with spack to think through when you'd do this (cross compiling??).

Reply:
Is the pattern `provides('foo', when="platform=bar")` supposed to work? 

When I tried this previously in https://github.com/spack/spack/pull/43002#issuecomment-2000345320 it didn't work. Trying it again now still doesn't. I tried the following permutations:

```
provides("uuid", when="platform=darwin")

for x in ('linux', 'cray', 'darwin'):
    with when(f"platform={x}"):
        provides("uuid")

with when(f"platform=darwin"):
     provides("uuid")
```

all fail with:

```
spack -d providers uuid
==> [2024-07-27-21:29:22.701764] Reading config from file /Users/cmarsh/Documents/science/code/spack/etc/spack/defaults/config.yaml
==> [2024-07-27-21:29:22.715155] Reading config from file /Users/cmarsh/.spack/config.yaml
==> [2024-07-27-21:29:22.718082] Imported providers from built-in commands
==> [2024-07-27-21:29:22.718338] Imported providers from built-in commands
==> [2024-07-27-21:29:22.718600] Reading config from file /Users/cmarsh/Documents/science/code/spack/etc/spack/defaults/repos.yaml
==> [2024-07-27-21:29:22.719690] Reading config from file /Users/cmarsh/.spack/repos.yaml
==> [2024-07-27-21:29:22.720291] Reading config from file /Users/cmarsh/Documents/science/code/spack/etc/spack/defaults/packages.yaml
==> [2024-07-27-21:29:22.742450] Reading config from file /Users/cmarsh/Documents/science/code/spack/etc/spack/defaults/darwin/packages.yaml
==> [2024-07-27-21:29:22.749395] Reading config from file /Users/cmarsh/.spack/packages.yaml
Traceback (most recent call last):
  File "/Users/cmarsh/Documents/science/code/spack/bin/spack", line 52, in <module>
    sys.exit(main())
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack_installable/main.py", line 42, in main
    sys.exit(spack.main.main(argv))
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/main.py", line 1070, in main
    return _main(argv)
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/main.py", line 1023, in _main
    return finish_parse_and_run(parser, cmd_name, args, env_format_error)
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/main.py", line 1053, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/main.py", line 650, in _invoke_command
    return_val = command(parser, args)
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/cmd/providers.py", line 27, in providers
    valid_virtuals = sorted(spack.repo.PATH.provider_index.providers.keys())
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/llnl/util/lang.py", line 754, in __getattr__
    return getattr(self.instance, name)
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/repo.py", line 791, in provider_index
    self._provider_index.merge(repo.provider_index)
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/repo.py", line 1139, in provider_index
    return self.index["providers"]
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/repo.py", line 620, in __getitem__
    self._build_all_indexes()
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/repo.py", line 633, in _build_all_indexes
    self.indexes[name] = self._build_index(name, indexer)
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/repo.py", line 665, in _build_index
    indexer.write(new)
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/repo.py", line 549, in write
    self.index.to_json(stream)
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/provider_index.py", line 170, in to_json
    provider_list = self._transform(
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/provider_index.py", line 71, in _transform
    return _transform(self.providers, transform_fun, out_mapping_type)
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/provider_index.py", line 271, in _transform
    return dict(
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/provider_index.py", line 272, in <genexpr>
    (name, out_mapping_type([transform_fun(vpkg, pset) for vpkg, pset in mapiter(mappings)]))
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/provider_index.py", line 272, in <listcomp>
    (name, out_mapping_type([transform_fun(vpkg, pset) for vpkg, pset in mapiter(mappings)]))
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/provider_index.py", line 171, in <lambda>
    lambda vpkg, pset: [vpkg.to_node_dict(), [p.to_node_dict() for p in pset]], list
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/provider_index.py", line 171, in <listcomp>
    lambda vpkg, pset: [vpkg.to_node_dict(), [p.to_node_dict() for p in pset]], list
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/spec.py", line 2125, in to_node_dict
    d.update(self.architecture.to_dict())
  File "/Users/cmarsh/Documents/science/code/spack/lib/spack/spack/spec.py", line 556, in to_dict
    ("target", self.target.to_dict_or_value()),
AttributeError: 'NoneType' object has no attribute 'to_dict_or_value'
```